macros/src/codegen.rs

```diff
@@ -355,60 +383,81 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
                             .zip(out_states.iter().zip(guard_action_parameters.iter().zip(guard_action_ref_parameters.iter()))),
                     )
                     .map(|(guard, (action, (out_state, (g_a_param, g_a_ref_param))))| {
-                        if let Some(AsyncIdent {ident: g, is_async: is_g_async}) = guard {
-                            let guard_await = match is_g_async {
-                                true => { sm_is_async = true; quote! { .await } },
-                                false => quote! { },
-                            };
-                            if let Some(AsyncIdent {ident: a, is_async: is_a_async}) = action {
-                                let action_await = match is_a_async {
-                                    true => { sm_is_async = true; quote! { .await } },
-                                    false => quote! { },
-                                };
-                                quote! {
-                                    let guard_result = self.context.#g(#temporary_context_call #g_a_ref_param) #guard_await;
-                                    self.context.log_guard(stringify!(#g), &guard_result);
-                                    if let Err(e) = guard_result {
-                                        self.state = Some(#states_type_name::#in_state);
-                                        return Err(#error_type_name::GuardFailed(e));
-                                    }
-                                    let _data = self.context.#a(#temporary_context_call #g_a_param) #action_await;
-                                    self.context.log_action(stringify!(#a));
-                                    let out_state = #states_type_name::#out_state;
-                                    self.context.log_state_change(&out_state);
-                                    self.state = Some(out_state);
+                        let streams: Vec<TokenStream> =
+                            guard.iter()
+                            .zip(action.iter().zip(out_state)).map(|(guard, (action,out_state))| {
+                                let (is_async,action_code) = generate_action(action, &temporary_context_call, g_a_param);
+                                if is_async {
+                                    sm_is_async = is_async;
```

Comment:
```suggestion
                                sm_is_async |= is_async;
```

nit: This could be simplified

Reply:
changed